### PR TITLE
Add <Route strict>

### PR DESCRIPTION
**Pull request by rafrex**
A fix for #4170 

Can now add a `strict` boolean prop to `Route` (if the strict prop is omitted it defaults to `true`):
```js
// this will match both /foo and /foo/
<Route exact path="/foo" strict={false} component={foo} />
```

I ran the tests and it passes all of them, and in my limited testing it works both inside and outside of `<Switch />`

### CODE REVIEW

**Comment by mjackson** [comment]
Thanks, @rafrex :) Right now our automated tests on v4 are not very complete, unfortunately. Would you mind booting the website and clicking through the examples to make sure everything still works?

**Comment by rafrex** [comment]
I clicked through all the examples and they seem to all work.

Note that I locally changed the default for `strict` to false in the router, and then booted the site so I  wouldn't have to add `strict={false}` to every route in the examples. I also tested a few routes by adding the `strict` prop as well.

**Comment by simonkberg** [comment]
Curious, how come `strict` is enabled by default? 
Not matching trailing slashes seems like the edge case here, which would be in line with the default behavior of `path-to-regex` (which mimics express' behavior).

**Comment by mjackson** [comment]
Sounds good, @rafrex. Can we get a few tests around this too? Ideally we'd have 2 tests: one for `strict = false` and another for `strict = true`.

**Comment by rafrex** [comment]
@simenbrekken I left the default as per how it worked previously. I agree it might make sense to change it to `false`.

**Comment by Phoenixmatrix** [comment]
Just a nitpicky 2 cents: the fact that the internal implementation (the path-to-regexp) uses an argument called "strict" is kind of an implementation detail, and that terminology isn't super intuitive for a user of the router. (like, wtf is "strict...oh, its a param for an internal dependency!".

if that gets exposed as a public api, it probably needs a better name.

**Comment by rafrex** [comment]
I actually don't mind the `strict` prop name because RRv4 doc says refer to `path-to-regex` to figure out how `path="..."` will match. But also don't have a strong preference for it either.

Maybe the default could be `false` and the prop could be `matchTrailingSlash` or something.

@mjackson what do think the default for `strict` should be? And what do you think the prop should be called? (I think both of these are high level api decisions).

**Comment by mjackson** [comment]
I think `strict` is a fine name and we should probably use the same default as `path-to-regexp` does. I'm sure those folks have good reasons by now for choosing the defaults they have.

**Comment by rafrex** [comment]
Change the default for strict to `false`.

I don't have experience with karma/mocha, someone with more experience want to write the tests?

**Comment by mjackson** [comment]
All you have to do is copy/paste some of the existing tests. Look at `integration-test.js` for hints.

**Comment by mjackson** [comment]
Also, we need to make sure we have tests for cases like https://github.com/ReactTraining/react-router/issues/4130 where two routes have similar URLs. @SimeonC created https://github.com/ReactTraining/react-router/pull/4133 with some tests that we may be able to use here (with attribution, ofc).

**Comment by SimeonC** [comment]
I'll just reply here - I can't update/create PRs currently as I'm without a PC for a month or two while moving so feel free to repurpose any of my PRs as needed.

**Comment by rafrex** [comment]
@mjackson I added tests for the strict prop with trailing slash.

I also added tests for similar paths (`/some` and `/some-path`). The problem in #4130 was with active links, which are no longer part of v4 (as far as I can tell), so I tested it using `children` and making sure the `match` boolean is correct (which seems to be how one would do active links now).

```js
<Route path="/some-path" children={({ match }) => (
  <h1>{match ? TEXT2 : TEXT1}</h1>
)}/>
```

**Comment by mjackson** [comment]
Thanks, @rafrex! Looks like there are some conflicts here, probably because I moved some stuff around when organizing the monorepo last week. Would you please rebase?

Also, sorry for all the churn. I'm trying to get things to settle down as quickly as I can.

**Comment by rafrex** [comment]
@mjackson I rebased, should be all set!

File: packages/react-router/modules/__tests__/integration-test.js [hunk 1]
```diff
@@ -38,6 +38,79 @@ describe('Integration Tests', () => {
     expect(div.innerHTML).toNotContain(TEXT)
   })
 
+  it('renders an exact route with trailing slash', () => {
```
Comment (pshrmn) [inline packages/react-router/modules/__tests__/integration-test.js]: Should there also be a test of the inverse? `<Route path='/somepath/'>` when `location.pathname = '/somepath'`?

Comment (rafrex) [inline packages/react-router/modules/__tests__/integration-test.js]: Sure, why not - added.

File: packages/react-router/modules/Core.js [hunk 2]
```diff
@@ -1,18 +1,18 @@
 import React, { PropTypes } from 'react'
 import pathToRegexp from 'path-to-regexp'
 
-const patternCache = { true: {}, false: {} }
+const patternCache = { truetrue: {}, truefalse: {}, falsefalse: {}, falsetrue: {} }
```
Comment (timdorr) [inline packages/react-router/modules/Core.js]: This is fine for now, but put a pin here to refactor this relative nonsensical code soon. We're bound to sprout a bug here eventually 🙃

**Comment by mjackson** [comment]
Thanks, @rafrex. I'll take it from here ;)
